Resharding: Online clone.

michael-berlin opened the pull request:
@alainjobart 

Only minor cleanups and the addition of the flags --online and --offline so far.

As always, I recommend to review it on a per commit basis.

Next steps I want to do are:
- add a `ResultReader` interface which will be implemented by `QueryResultReader` and a future `ResultMerger`
- extend `clone()` such that it puts M sourceReaders and N destReaders into a `ResultMerger` respectively; both `ResultMerger` will feed their output into `RowDiffer2`
- let `RowDiffer2` emit SQL statements to fix any found diffs
- implement `ResultMerger`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/1796)
<!-- Reviewable:end -->

alainjobart commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 13 of 13 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[go/vt/worker/split_clone.go, line 35 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEPQHVZCQBedCLIt_l:-KKEPQHVZCQBedCLIt_m:64542206) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L35)):*
> ```Go
> )
> 
> type clonePhase string
> ```

let's just use the other constants here, WorkerStateCloneOnline and WorkerStateCloneOffline.

---

*[go/vt/worker/split_clone.go, line 386 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEOHYCuzrOLRXLECIv:-KKEOHYCuzrOLRXLECIw:-216921073) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L386)):*
> ```Go
> // - mark it as 'worker' pointing back to us
> // - get the aliases of all the source tablets
> func (scw *SplitCloneWorker) findSourceTablets(ctx context.Context) error {
> ```

Change the method name to match comment

---

*[go/vt/worker/split_clone.go, line 507 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEOXy2JQLblLBl7lTa:-KKEOXy2JQLblLBl7lTb:-1348298204) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L507)):*
> ```Go
> // Assumes that the schema has already been created on each destination tablet
> // (probably from vtctl's CopySchemaShard)
> func (scw *SplitCloneWorker) clone(ctx context.Context, clonePhase clonePhase) error {
> ```

why not pass in the WorkerStateCloneOnline / Offline constants and use these, instead of declaring a new one?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796#-:-KKEPtM305aawgFYz8zF:1417088721)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
Review status: 1 of 14 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[go/vt/worker/split_clone.go, line 35 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEPQHVZCQBedCLIt_l:-KKEpX2p_5DoHfCCWRjV:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L35)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> let's just use the other constants here, WorkerStateCloneOnline and WorkerStateCloneOffline.
</details>
Done.

---

*[go/vt/worker/split_clone.go, line 386 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEOHYCuzrOLRXLECIv:-KKEpg9dZSzGwgVHFLvY:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L386)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> Change the method name to match comment
</details>
Done.

---

*[go/vt/worker/split_clone.go, line 507 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEOXy2JQLblLBl7lTa:-KKEph3BBa9RnsAo2sGJ:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L507)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> why not pass in the WorkerStateCloneOnline / Offline constants and use these, instead of declaring a new one?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
@alainjobart Can you PTAL?

I did the following items from the TODO list above:

- add a ResultReader interface which will be implemented by QueryResultReader and a future ResultMerger
- extend clone() such that it puts M sourceReaders and N destReaders into a ResultMerger respectively; both ResultMerger will feed their output into RowDiffer2

alainjobart commented:
Couple minor comments, LGTM so far.

michael-berlin commented:
@alainjobart PTAL

Offline phase for empty destinations is working now.

What's left is the implementation of the ResultMerger and writing tests for the repair cases (update, delete).

---

Review status: 0 of 19 files reviewed at latest revision, 3 unresolved discussions.

---

*[go/vt/worker/split_clone.go, line 35 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEPQHVZCQBedCLIt_l:-KKHvz00uaA4hezdrvBQ:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L35)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Done.
</details>
Done.

---

*[go/vt/worker/split_clone.go, line 386 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEOHYCuzrOLRXLECIv:-KKHw2esdRzn7HSyj62p:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L386)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Done.
</details>
Done.

---

*[go/vt/worker/split_clone.go, line 507 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKEOXy2JQLblLBl7lTa:-KKHw3lgC3L7gtTtuGJ1:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1114b0987a24576e872aee348e03160e292db2a5/go/vt/worker/split_clone.go#L507)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Done.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796#-:-KKHvCgqWHJCFEmdtWII:286595217)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> so far.

Let's add unit test cases for the different pipeline stages once the interfaces and implementations are stable.

Overall, the main comments are:
- logging of differences in RowDiffer2 is not necessary.
- separating each class in its own file may be better.
- using the context everywhere for cancellation might be better.

Right now, if one streaming query fails, we abort the entire process. It is fine short term, but it would be somewhat easy to restart that query where it stopped. We have the last value that was read on it, so issuing a query to read from there wouldn't be too hard (I know it stopped on a random value, so we'd need a weird 'WHERE' clause for the starting point when the PK has multiple rows, but it's doable). Then we can go to any rdonly guy again, using the same health check.

---

Reviewed 5 of 9 files at r5.
Review status: 14 of 19 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[go/vt/worker/row_differ.go, line 219 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJRhLh29YGwMUoXzXI:-KKJRhLh29YGwMUoXzXJ:1495463402) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L219)):*
> ```Go
> 		// they're the same. Logging a regular difference
> 		// then, and advancing both.
> 		// TODO(mberlin): Does this case happen in practice?
> ```

It should right? It's the case where the contents of a row have changed...

---

*[go/vt/worker/row_differ.go, line 220 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJRm3t1UjSnm5JFKfA:-KKJRm3t1UjSnm5JFKfB:1790073770) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L220)):*
> ```Go
> 		// then, and advancing both.
> 		// TODO(mberlin): Does this case happen in practice?
> 		if dr.mismatchedRows < 10 {
> ```

let's remove the logging of the first 10 rows, it's useless when diff is expected.

---

*[go/vt/worker/row_differ.go, line 257 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSv0mfdfUntXJmBQL:-KKJSv0mfdfUntXJmBQM:-1752962579) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L257)):*
> ```Go
> }
> 
> // RowRouter allows to find out which shard's key range contains a given
> ```

How about putting this in its own file?

---

*[go/vt/worker/row_differ.go, line 290 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSqdKu0zFwwB6AaKE:-KKJSqdKu0zFwwB6AaKF:-1752962579) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L290)):*
> ```Go
> }
> 
> // RowAggregator aggregates SQL repair statements into one statement.
> ```

How about putting this in its own file?

---

*[go/vt/worker/row_differ.go, line 322 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSXiKnYxHbCDn5FLQ:-KKJSXiKnYxHbCDn5FLR:-901263456) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L322)):*
> ```Go
> 
> // NewRowAggregator returns a RowAggregator.
> func NewRowAggregator(limit int, keyspaceAndShard string, insertChannel chan string, abort <-chan struct{}, dbName string, td *tabletmanagerdatapb.TableDefinition, repairType repairType) *RowAggregator {
> ```

We should use a context and not an abort channel?

---

*[go/vt/worker/row_differ.go, line 325 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSbmoPMpkqZ2QQrdC:-KKJSbmpJc0iJ2Hu_-9X:-1711793772) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L325)):*
> ```Go
> 	// Construct head and tail base commands for the repair statement.
> 	var baseCmdHead, baseCmdTail string
> 	switch repairType {
> ```

the two methods that do something here are both a switch on the type. Would it be easier to have a base class and 3 derived class that each do insert, delete or update?

---

*[go/vt/worker/row_differ.go, line 433 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJUEWy6-zIcSBMnkbe:-KKJUEWy6-zIcSBMnkbf:1995132972) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L433)):*
> ```Go
> }
> 
> // Close fluses any pending aggregates which haven't been sent out yet.
> ```

fluses -> flushes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796#-:-KKJUe1Tcj8YyxtYQlr4:-46031471)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
Reviewed 1 of 6 files at r4, 4 of 9 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796)*
<!-- Sent from Reviewable.io -->

enisoc commented:
Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 6 of 13 files at r3, 1 of 6 files at r4, 2 of 9 files at r5, 17 of 20 files at r6, 6 of 6 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[go/vt/worker/result_merger.go, line 20 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKker131iTvurDIYi5z:-KKker131iTvurDIYi6-:1178680252) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L20)):*
> ```Go
> // The current value of 64 rows tries to aim at a total size of 4k bytes
> // for the returned Result (i.e. we assume an average row size of 64 bytes).
> const ResultSizeRows = 64
> ```

This seems like the kind of thing you'd want as a flag so you don't have to make a new build to try a different value.

---

*[go/vt/worker/result_merger.go, line 32 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoPumptHw6G_O7v7qq:-KKoPumptHw6G_O7v7qr:-1387020573) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L32)):*
> ```Go
> 	output [][]sqltypes.Value
> 	// currentIndex is the index of the next row which will be set in "output".
> 	currentIndex int
> ```

Why not use `len(output)` as the `currentIndex` and just `append()`?

---

*[go/vt/worker/result_merger.go, line 38 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKkfVikW4tHpqZIZAdl:-KKkfVikW4tHpqZIZAdm:-194206789) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L38)):*
> ```Go
> 	// lastRowReaderDrained becomes true when there is only one input ResultReader
> 	// left and its respective RowReader stopped after the current Result. From
> 	// there on, we can switch to an "endgame" mode where we can read the Result
> ```

Do you have evidence or an argument that this endgame mode is likely to be triggered, and will be a significant benefit? It would be conceptually simpler if all this processing was in terms of rows as much as possible (RowMerger), unless there's a convincing argument that it's inefficient. For example, the pipeline would look like: []ResultReader -> []RowReader -> RowMerger -> RowCombiner -> ResultReader. You're already doing this pipeline, but you're combining some steps into one, when they could be orthogonal.

---

*[go/vt/worker/result_merger.go, line 147 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoRi1tjCy2mQz0Ogex:-KKoRi1tjCy2mQz0Ogey:-1932274089) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L147)):*
> ```Go
> 
> 	if rm.currentIndex < len(rm.output) {
> 		// Output is incomplete because it's the very last one. Adjust its length.
> ```

If you use `append()` you won't need to do this.

---

*[go/vt/worker/result_merger.go, line 163 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoSJwoOyVtaCn0h_Qq:-KKoSJwoOyVtaCn0h_Qr:-1891523256) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L163)):*
> ```Go
> func (rm *ResultMerger) deleteInput(deleteMe ResultReader) {
> 	for i, input := range rm.inputs {
> 		if input == deleteMe {
> ```

This assumes everything that implements `ResultReader` uses a pointer receiver. Make sure that's mentioned as a requirement in the doc comment for `ResultReader`.

---

*[go/vt/worker/row_aggregator.go, line 33 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKp8iOiyo2yQvSjvY60:-KKp8iOiyo2yQvSjvY61:2096634693) ([raw file](https://github.com/youtube/vitess/blob/1799e5d3ce6a906a34de1593463174805a03f09d/go/vt/worker/row_aggregator.go#L33)):*
> ```Go
> 	maxRows int
> 	maxSize int
> 	// keyspaceAndShard is only used for error messages.
> ```

A string that's passed in only for error messages feels like a broken abstraction layer. If this layer doesn't need to know the keyspace/shard to do its actual job, why can't the layer above inject keyspace/shard context into errors?

---

*[go/vt/worker/row_aggregator.go, line 113 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKpCIvyamOfgvjq0qCT:-KKpCIvyamOfgvjq0qCU:763156338) ([raw file](https://github.com/youtube/vitess/blob/1799e5d3ce6a906a34de1593463174805a03f09d/go/vt/worker/row_aggregator.go#L113)):*
> ```Go
> // flush sends out the current aggregation buffer. It returns true when the
> // pipeline was aborted.
> func (ra *RowAggregator) flush() bool {
> ```

Why is abort signaled as a bool, rather than returning an error?

---

*[go/vt/worker/row_differ.go, line 219 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJRhLh29YGwMUoXzXI:-KKp1ovHpBzKRXQIOwof:-1397761957) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L219)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> It should right? It's the case where the contents of a row have changed...
</details>
We would only get here if `RowsEqual()` says the primary keys are different, but `CompareRows` says the primary keys are equal. The difference is that one compares raw values and the other parsed values. In other words, the raw values returned by MySQL were different but they became equal after we parsed them into ints/floats (due to leading/trailing zeros, for example). So this can happen if MySQL is inconsistent in how it prints a given number.

---

*[go/vt/worker/row_differ.go, line 220 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJRm3t1UjSnm5JFKfA:-KKp465zPQdOO5VsMp5O:5172523) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L220)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> let's remove the logging of the first 10 rows, it's useless when diff is expected.
</details>
+1

---

*[go/vt/worker/row_differ.go, line 257 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSv0mfdfUntXJmBQL:-KKp4JxAEUflrapx7ULd:1399128167) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L257)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> How about putting this in its own file?
</details>
RowRouter seems actually too simple as it is. It could just be kept as a `route()` method on `RowDiffer2` for now, and then moved into `RowUpdater` when that gets split off from `RowDiffer2`.

---

*[go/vt/worker/row_differ.go, line 104 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoo4FzUU7loNUORJ1h:-KKoo4FzUU7loNUORJ1i:-678620753) ([raw file](https://github.com/youtube/vitess/blob/a1af78df545de69f9f4cfe57ab28cd906029e2d7/go/vt/worker/row_differ.go#L104)):*
> ```Go
> // Go runs the diff and repair.
> // If an error occurs, it will return and stop.
> func (rd *RowDiffer2) Go(log logutil.Logger) (DiffReport, error) {
> ```

A method named `Go` in Go is usually expected to launch a goroutine. I would prefer something like `Run()` to avoid such connotations. Or maybe since the struct is named `*Differ` it should be`Diff()`.

---

*[go/vt/worker/row_differ.go, line 29 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKop0-JYWYYw-6-sTaR:-KKop0-JYWYYw-6-sTaS:-1868366313) ([raw file](https://github.com/youtube/vitess/blob/1799e5d3ce6a906a34de1593463174805a03f09d/go/vt/worker/row_differ.go#L29)):*
> ```Go
> 	update
> 	// delet is misspelled because "delete" is a reserved word.
> 	delet
> ```

Keep in mind these constants are package-scoped, not file-scoped. Does the meaning of `insert` as being a `repairType` make sense across all of `worker`? I would call these something like `diffInsert` to hint that they belong to the diff module within the worker package.

---

*[go/vt/worker/row_differ.go, line 225 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoqfTXjk21S2q4fJls:-KKoqfTXjk21S2q4fJlt:-2045633651) ([raw file](https://github.com/youtube/vitess/blob/1799e5d3ce6a906a34de1593463174805a03f09d/go/vt/worker/row_differ.go#L225)):*
> ```Go
> 	}
> 
> 	// Close and flush all aggregators in case they have data buffered.
> ```

Should aggregators not be closed if we return early due to an error?

---

*[go/vt/worker/tablet_tracker.go, line 59 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKkaMnVRAI4rZzGgXXZ:-KKkaMnVRAI4rZzGgXX_:1262864150) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/tablet_tracker.go#L59)):*
> ```Go
> 				count := t.usedTablets[key]
> 				count++
> 				t.usedTablets[key] = count
> ```

I think you can just `t.usedTablets[key]++`

---

*[go/vt/worker/tablet_tracker.go, line 96 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKkWRGDnc9TrXN1Oknx:-KKkWRGDnc9TrXN1Okny:1968421127) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/tablet_tracker.go#L96)):*
> ```Go
> // Sort by value taken from: https://groups.google.com/d/msg/golang-nuts/FT7cjmcL7gw/Gj4_aEsE_IsJ
> 
> // Pair represents a table (Key) and its usage (Value).
> ```

s/table/tablet/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
I've addressed all comments.

enisoc@ can you please have a look at the diff since your last review? Thanks!

---

Review status: 18 of 38 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[go/vt/worker/result_merger.go, line 20 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKker131iTvurDIYi5z:-KKtYZGQrV7g210KalLJ:-1894045353) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L20)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> This seems like the kind of thing you'd want as a flag so you don't have to make a new build to try a different value.
</details>
I'll keep it a constant for now and turn this class later from a ResultMerger into a RowMerger. Then we won't need this anymore.

---

*[go/vt/worker/result_merger.go, line 32 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoPumptHw6G_O7v7qq:-KKuJtRwJWxVz9RsEHtU:-499146009) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L32)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Why not use `len(output)` as the `currentIndex` and just `append()`?
</details>
Done.

---

*[go/vt/worker/result_merger.go, line 38 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKkfVikW4tHpqZIZAdl:-KKuJuRAvhAg5Q7CGfnI:-89210996) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L38)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Do you have evidence or an argument that this endgame mode is likely to be triggered, and will be a significant benefit? It would be conceptually simpler if all this processing was in terms of rows as much as possible (RowMerger), unless there's a convincing argument that it's inefficient. For example, the pipeline would look like: []ResultReader -> []RowReader -> RowMerger -> RowCombiner -> ResultReader. You're already doing this pipeline, but you're combining some steps into one, when they could be orthogonal.
</details>
No, we don't have any evidence.

As discussed offline I'll refactor this code and turn the ResultMerger into a RowMerger. This will simplify a lot of things.

---

*[go/vt/worker/result_merger.go, line 147 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoRi1tjCy2mQz0Ogex:-KKuK1UpoPA8vSpUu7Kt:-499146009) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L147)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> If you use `append()` you won't need to do this.
</details>
Done.

---

*[go/vt/worker/result_merger.go, line 163 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoSJwoOyVtaCn0h_Qq:-KKuM-V-h5GTpNAxFGAM:-499146009) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/result_merger.go#L163)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> This assumes everything that implements `ResultReader` uses a pointer receiver. Make sure that's mentioned as a requirement in the doc comment for `ResultReader`.
</details>
Done.

---

*[go/vt/worker/row_aggregator.go, line 33 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKp8iOiyo2yQvSjvY60:-KKuMDjJvza2IA6qvjrD:611234493) ([raw file](https://github.com/youtube/vitess/blob/1799e5d3ce6a906a34de1593463174805a03f09d/go/vt/worker/row_aggregator.go#L33)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> A string that's passed in only for error messages feels like a broken abstraction layer. If this layer doesn't need to know the keyspace/shard to do its actual job, why can't the layer above inject keyspace/shard context into errors?
</details>
I've deleted this field for now and omitted this information from the error messages. We can add it back later at a higher level when we move the RowAggregator instances from the RowDiffer struct to a future RowReconciler struct.

---

*[go/vt/worker/row_aggregator.go, line 113 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKpCIvyamOfgvjq0qCT:-KKuOz2eCXyaZ7b3Bz7_:1064456251) ([raw file](https://github.com/youtube/vitess/blob/1799e5d3ce6a906a34de1593463174805a03f09d/go/vt/worker/row_aggregator.go#L113)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Why is abort signaled as a bool, rather than returning an error?
</details>
Done.

Changed to return an error.

---

*[go/vt/worker/row_differ.go, line 219 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJRhLh29YGwMUoXzXI:-KKtWrU2GbTO2AAJmc6R:1573799968) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L219)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> We would only get here if `RowsEqual()` says the primary keys are different, but `CompareRows` says the primary keys are equal. The difference is that one compares raw values and the other parsed values. In other words, the raw values returned by MySQL were different but they became equal after we parsed them into ints/floats (due to leading/trailing zeros, for example). So this can happen if MySQL is inconsistent in how it prints a given number.
</details>
Done.

I've removed the TODO and added an explanation.

---

*[go/vt/worker/row_differ.go, line 220 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJRm3t1UjSnm5JFKfA:-KKtXZe6dDIU1AvTaJSq:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L220)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> +1
</details>
Done.

---

*[go/vt/worker/row_differ.go, line 257 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSv0mfdfUntXJmBQL:-KKuXK12lJB-GGyNN8hE:550898135) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L257)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> RowRouter seems actually too simple as it is. It could just be kept as a `route()` method on `RowDiffer2` for now, and then moved into `RowUpdater` when that gets split off from `RowDiffer2`.
</details>
I'll do this refactor in a separate pull request later.

---

*[go/vt/worker/row_differ.go, line 290 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSqdKu0zFwwB6AaKE:-KKuY9GLRkYPbTiUxsgK:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L290)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> How about putting this in its own file?
</details>
Done.

---

*[go/vt/worker/row_differ.go, line 322 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSXiKnYxHbCDn5FLQ:-KKuYAc1ivCzSrtWlcti:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L322)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> We should use a context and not an abort channel?
</details>
Done.

---

*[go/vt/worker/row_differ.go, line 325 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJSbmoPMpkqZ2QQrdC:-KKuYBTXkQ1k48yCiiRW:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L325)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> the two methods that do something here are both a switch on the type. Would it be easier to have a base class and 3 derived class that each do insert, delete or update?
</details>
Done.

---

*[go/vt/worker/row_differ.go, line 433 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKJUEWy6-zIcSBMnkbe:-KKuXbbn_96aUPcqns1x:-499146009) ([raw file](https://github.com/youtube/vitess/blob/1502504b9f1435bf4fa4347aef15e68382f24c2a/go/vt/worker/row_differ.go#L433)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> fluses -> flushes
</details>
Done.

---

*[go/vt/worker/row_differ.go, line 104 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoo4FzUU7loNUORJ1h:-KKtYLufDEj9U3P7dVUI:1094104368) ([raw file](https://github.com/youtube/vitess/blob/a1af78df545de69f9f4cfe57ab28cd906029e2d7/go/vt/worker/row_differ.go#L104)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> A method named `Go` in Go is usually expected to launch a goroutine. I would prefer something like `Run()` to avoid such connotations. Or maybe since the struct is named `*Differ` it should be`Diff()`.
</details>
Done.

Renamed to `Diff()`.

---

*[go/vt/worker/row_differ.go, line 29 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKop0-JYWYYw-6-sTaR:-KKuSri08DUEl6qC0C5X:-2084107750) ([raw file](https://github.com/youtube/vitess/blob/1799e5d3ce6a906a34de1593463174805a03f09d/go/vt/worker/row_differ.go#L29)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Keep in mind these constants are package-scoped, not file-scoped. Does the meaning of `insert` as being a `repairType` make sense across all of `worker`? I would call these something like `diffInsert` to hint that they belong to the diff module within the worker package.
</details>
Done.

Yes, I'm aware that they are package scoped. I've renamed the type to `DiffType` and changed the variable names as well:

```go
	DiffMissing DiffType = iota
	DiffNotEqual
	DiffExtraneous
```

---

*[go/vt/worker/row_differ.go, line 225 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKoqfTXjk21S2q4fJls:-KKuXC3C26ZVGQj2afZm:59053296) ([raw file](https://github.com/youtube/vitess/blob/1799e5d3ce6a906a34de1593463174805a03f09d/go/vt/worker/row_differ.go#L225)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Should aggregators not be closed if we return early due to an error?
</details>
Done.

No, they don't have to be closed. I've renamed Close() to Flush() to clear up expectations here.

---

*[go/vt/worker/tablet_tracker.go, line 59 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKkaMnVRAI4rZzGgXXZ:-KKuY-lmYslfsvznVtVh:-499146009) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/tablet_tracker.go#L59)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> I think you can just `t.usedTablets[key]++`
</details>
Done.

---

*[go/vt/worker/tablet_tracker.go, line 96 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKkWRGDnc9TrXN1Oknx:-KKuY6Ao4p-Okh4TB2kg:-499146009) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/go/vt/worker/tablet_tracker.go#L96)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> s/table/tablet/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796#-:-KKuYzlrysHl2qiWac8l:1225693141)*
<!-- Sent from Reviewable.io -->

enisoc commented:
Reviewed 20 of 20 files at r9.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[go/vt/worker/row_aggregator.go, line 41 \[r9\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKv26EFfwvwyt9foaCj:-KKv26EFfwvwyt9foaCk:151191516) ([raw file](https://github.com/youtube/vitess/blob/2c9f640a1d012a8316d51c24f6f2498b08ea60b2/go/vt/worker/row_aggregator.go#L41)):*
> ```Go
> 	builder       QueryBuilder
> 	// statsCounters has Counters to track how many rows were changed per
> 	// diffTyp.
> ```

Lost the 'e' on DiffType?

---

*[test/resharding.py, line 192 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKv5x8xBDeFqov-DDgr:-KKv5x8yyQPPlmW2TifL:1002688979) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/test/resharding.py#L192)):*
> ```Python
> id bigint not null,
> msg varchar(64),
> custom_sharding_key ''' + t + ''' not null,
> ```

`custom_sharding_key` was there to make sure sharding works with a `sharding_column_name` other than `keyspace_id`. That was a bug before, so this avoid regressions. If you want to make them all the same so they can share code, it would be best to make them all something other than `keyspace_id` since that's the most likely regression.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
@enisoc PTAL 

This PR is ready to be merged now.

Changes this your last review:
- addressed your comments
- added old SplitClone back as `LegacySplitClone`
- fixed an issue where SplitClone failed and the throttlers were not properly closed

---

Review status: 26 of 45 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[go/vt/worker/row_aggregator.go, line 41 \[r9\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKv26EFfwvwyt9foaCj:-KL4F7Q9iqDic7gLYqQH:-499146009) ([raw file](https://github.com/youtube/vitess/blob/2c9f640a1d012a8316d51c24f6f2498b08ea60b2/go/vt/worker/row_aggregator.go#L41)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Lost the 'e' on DiffType?
</details>
Done.

---

*[test/resharding.py, line 192 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1796#-KKv5x8xBDeFqov-DDgr:-KL4tiJCslZObzfcF16o:43206545) ([raw file](https://github.com/youtube/vitess/blob/f7c3db8e72c59b280f34fc241e8347b7fa1daa5e/test/resharding.py#L192)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> `custom_sharding_key` was there to make sure sharding works with a `sharding_column_name` other than `keyspace_id`. That was a bug before, so this avoid regressions. If you want to make them all the same so they can share code, it would be best to make them all something other than `keyspace_id` since that's the most likely regression.
</details>
Done.

I've renamed the column from `keyspace_id` to `custom_ksid_col`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796#-:-KL4u0mMT-KpKDRP2sTf:-1875372039)*
<!-- Sent from Reviewable.io -->

enisoc commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 20 of 21 files at r10, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1796#-:-KLFPwCLQ06KLrPk59DA:1417088721)*
<!-- Sent from Reviewable.io -->
